lib/pbench/server/database/models/datasets.py

```diff
@@ -450,14 +431,16 @@ def add(self):
         try:
             Database.db_session.add(self)
             Database.db_session.commit()
-        except IntegrityError as e:
-            Database.db_session.rollback()
-            self.logger.warning("Duplicate dataset {}: {}", self.name, e)
-            raise DatasetDuplicate(self.name) from None
         except Exception as e:
             Database.db_session.rollback()
-            self.logger.error("Can't add {} to DB: {}", str(self), str(e))
-            raise DatasetSqlError("adding", name=self.name) from e
+            raise decode_sql_error(
+                e,
+                on_duplicate=DatasetDuplicate,
+                on_null=DatasetSqlError,
+                fallback=DatasetSqlError,
+                operation="add",
+                dataset=self,
+            )
```

Comment:
Should there be a `from` clause at line 443?

Ditto line 460?

Reply:
Interestingly, the old code suppressed chaining on duplicate, and both cases are now collapsed. I'm not sure it matters since the cause is pretty obvious from context, but I guess I might as well try to be "pure". 😦

Reply:
It's not really a matter of purity:  without the `from`, the traceback will report, "During handling of the above exception, another exception occurred", which implies that there is a fault in the exception handling code; what we want is to cause a different exception to be raised, instead, so the question is whether the `from` clause should specify the original exception (so that the traceback reports the original exception as "the direct cause of" the new exception) or whether the `from` clause should specify `None` (so that the original exception is completely hidden) -- either way, there should be a `from`....

Reply:
> It's not really a matter of purity: without the `from`, the traceback will report, "During handling of the above exception, another exception occurred", which implies that there is a fault in the exception handling code; what we want is to cause a different exception to be raised, instead, so the question is whether the `from` clause should specify the original exception (so that the traceback reports the original exception as "the direct cause of" the new exception) or whether the `from` clause should specify `None` (so that the original exception is completely hidden) -- either way, there should be a `from`....

No, the point is more that we don't deal with traceback except when something ends up in a "server internal error" or debugging unit tests. If we were exposing direct Python APIs I'd be a lot more concerned about confusing external clients with "new exception" vs "cause" ... I'm not really worried that it'll confuse us; so while you're not wrong, I consider this more an issue of "language semantics purity" than of "pragmatic usability".